### PR TITLE
[Draft] 5 axis CNC kinematics (swivel head / rotary table), G43.4 RTCP, G10 Tool offsets

### PR DESCRIPTION
**Pull request by DerAndere1**
### Description

Do not merge. Do not test without adding the missing parts. This is an untested, unfinished draft of multi-axis kinematics that might allow true continuous 5-axis machining with Tool Center Point Control.

- G10 L2/L20 P_ X_ Y_ Z_: Set workspace offset like LinuxCNC (this was authored by Domush)
- G43.4: Enable Rotational Tool Center Point Control Mode (otherwise known as G43.4 RTCP, G234 TCPM, TCPC, TRAORI, TRAFO)
- G43: Enable simple tool length compensation
- G49: Disable Tool length compensation (cancels G43.4 and G43) for direct control of motors in joint-space
- G10 L1/L10 P_ X_ Y_ Z_: Set Tool Offsets (Tool length)

Missing:
- forward kinematics and proper reporting of positions

### Requirements

- Do not use this on hardware.
- 4 axis CNC machine with swivel head or 
- 5 axis CNC machine with swivel head / rotary table configuration (tilting head is B axis, horizontal rotary table is C axis
- Enable new options in Configuration: `XYZBC_HEAD_TABLE` and related new options.
- Based on PR https://github.com/MarlinFirmware/Marlin/pull/24334
- TOOLS must equal the number of non-extruder tools plus EXTRUDERS
- Currently best used with `SWITCHING_TOOLHEAD`, extruders/hotends as tools 0-7 followed by non-extruder tools, `HOTEND_OFFSET_[X,Y,Z]` set for all tools including hotends. Requires workspace offsets (`NO_WORKSPACE_OFFSETS` must be disabled).  

 
### Benefits

- See what it needs to accomplish this.
- Test previous software design choices
- up to 128 tools, including up to 8 extruders

### Configurations

[https://github.com/DerAndere1/Marlin/tree/Domush/config/examples/DerAndere](https://github.com/DerAndere1/Marlin/tree/Domush/config/examples/DerAndere)

### Related Issues

None

### CODE REVIEW

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -987,6 +987,21 @@
...
+// This machine has a swivel head (B axis parallel to the Y axis) and a horizontal rotary
+// table (C axis parallel to the Z axis).
+//#define XYZBC_HEAD_TABLE
+#if ENABLED(XYZBC_HEAD_TABLE)
+  #define TOOLS 1 // Number of tools;
+  // Machine rotary zero point offset is the distance between the center of rotation of the B axis to the gauge line at the tool head.
+  #define DEFAULT_MRZP_OFFSET_MM 100.0 // (mm)
+  #define DEFAULT_TOOL_OFFSET_X { 0.0 } // (mm) Tool offset X
+  #define DEFAULT_TOOL_OFFSET_Y { 0.0 } // (mm) Tool offset Y
+  #define DEFAULT_TOOL_OFFSET_Z { 0.0 } // (mm) Gauge length (Tool length)
```
Comment (GMagician) [inline Marlin/Configuration.h]: Maybe tool related offset should be variable. I'm thinking at manual tool change...

Comment (DerAndere1) [reply Marlin/Configuration.h]: it already is, see variable `tool_offsets[TOOLS]` which is initialized with these default values and which can be altered with G10 L1 or G10 L10. Tool length compensation is off by default and is only enabled with G43 (for 3 axis machining) or with G43.4 (RTCP for continuous 5 axis operation).

Comment (DerAndere1) [reply Marlin/Configuration.h]: In the long run, hotend offsets could be replaced by these, but that requires quite some refactoring to separate the concept of extruders and hotends / tools

File: Marlin/src/lcd/language/language_it.h [hunk 2]
```diff
@@ -191,8 +191,8 @@ namespace Language_it {
   LSTR MSG_IDEX_MODE_MIRRORED_COPY        = _UxGT("Copia speculare");
   LSTR MSG_IDEX_MODE_FULL_CTRL            = _UxGT("Pieno controllo");
   LSTR MSG_IDEX_DUPE_GAP                  = _UxGT("X-Gap-X duplicato");
-  LSTR MSG_HOTEND_OFFSET_Z                = _UxGT("2° ugello Z");
-  LSTR MSG_HOTEND_OFFSET_A                = _UxGT("2° ugello @");
+  LSTR MSG_HOTEND_OFFSET_Z                = _UxGT("2") LCD_STR_DEGREE _UxGT(" ugello Z");
```
Comment (GMagician) [inline Marlin/src/lcd/language/language_it.h]: ° is used as short form of second 'nd', not degree

Comment (DerAndere1) [reply Marlin/src/lcd/language/language_it.h]: fixed

**Comment by DerAndere1** [comment]
I will close this PR for now. It works for me but this should be divided into smaller PRs. This will stay available as part of my fork.

**Comment by DerAndere1** [comment]
To clarify the license status: The xyzbc_head_table.cpp kinematics implementation is based on a relicensed version of maxkins.c from LinuxCNC. Here is a copy of the E-mail conversation between the original author of maxkins.c (Chris Radek), and me:

```
Please find attached a relicensed version of that file for your use
under the terms of the GPL2 or GPL3 at your option.

Please be aware that I cannot relicense any of the files it
includes, but I am pretty sure you won't want those parts anyway.

Best regards,
Chris

> Dear Mr. Radek,
>
> I found your email address in the version control history of the 
> LinuxCNC software. I greatly appreciate your work on the maxkins CNC 
> kinematics. I am contributing to the open source 3D printer firmware 
> MarlinFirmware/Marlin.
>
> Would you agree to license a copy of your maxkins.c code [1] under the 
> General Public License 3.0 (GPL 3.0) ? The original maxkins.c file in 
> LinuxCNC can stay licensed under GPL 2.0.
>
> The difference between the license versions is only subtle [2] but the 
> newer license would be compatible with our project. Thanks!
>
>
> Links:
>
> [1] 
> https://github.com/LinuxCNC/linuxcnc/blob/ada63a4b3af8de6c16d59915502f6dec2155b81c/src/emc/kinematics/maxkins.c 
>
>
> [2] 
> https://stackoverflow.com/questions/41460/what-are-the-differences-between-gpl-v2-and-gpl-v3-licenses
>
>
> Kind regards
```
[maxkins.zip](https://github.com/MarlinFirmware/Marlin/files/9418388/maxkins.zip)

**Pull request closed without merging**